File: NinePSharp.Messages.FSharp/messages/Rxattrwalk.fs
namespace NinePSharp.Messages.FSharp

open System
open System.Buffers.Binary
open NinePSharp.Constants
open NinePSharp.Interfaces
open NinePSharp.Messages
open NinePSharp.Messages.FSharp.Protocol

type Rxattrwalk =
    { Size: uint32
      Tag: uint16
      XattrSize: uint64 }

    interface ISerializable with
        member this.Size = this.Size
        member this.Type = MessageTypes.Rxattrwalk
        member this.Tag = this.Tag
        member this.WriteTo(span: Span<byte>) =
            let mutable offset = 0
            writeHeaders this.Size this.Tag MessageTypes.Rxattrwalk span &offset
            BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(offset, 8), this.XattrSize)

    static member Parse (data: ReadOnlySpan<byte>) =
        let mutable offset = 0
        let size, tag = readHeaders data &offset
        let xattrSize = BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(offset, 8))
        { Size = size; Tag = tag; XattrSize = xattrSize }

File: NinePSharp/messages/Rxattrwalk.cs
namespace NinePSharp.Messages;

using System;
using System.Buffers.Binary;
using NinePSharp.Constants;
using NinePSharp.Interfaces;
using NinePSharp.Protocol;

public readonly struct Rxattrwalk : ISerializable
{
    public uint Size { get; }
    public MessageTypes Type => MessageTypes.Rxattrwalk;
    public ushort Tag { get; }
    public ulong XattrSize { get; }

    public Rxattrwalk(uint size, ushort tag, ulong xattrSize)
    {
        Size = size;
        Tag = tag;
        XattrSize = xattrSize;
    }

    public Rxattrwalk(ReadOnlySpan<byte> data)
    {
        Size = BinaryPrimitives.ReadUInt32LittleEndian(data[..4]);
        Tag = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(5, 2));
        int offset = NinePConstants.HeaderSize;
        XattrSize = BinaryPrimitives.ReadUInt64LittleEndian(data.Slice(offset, 8));
    }

    public void WriteTo(Span<byte> span)
    {
        span.WriteHeaders(Size, Tag, MessageTypes.Rxattrwalk);
        int offset = NinePConstants.HeaderSize;
        BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(offset, 8), XattrSize);
    }
}
